newsrevenuehub/rev-engine
# [DEV-699] Introduce roles and use them to programmatically manage access to orgs, rev programs, and other data models

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

NA

#### What's this PR do?

- Introduces role-based access for the primary resources exposed through the REST API.
- Introduces three new user roles: 
  - Hub admin - prototypically can CRUD everything
  - Org admin - prototypically can CRUD only things belonging to their org
  - Revenue program admin - prototypically can CRUD only things belonging to their RPs
- Allows role assignments to be managed in Django admin. Role assignments connect a user, a role type, an organization, and a set of revenue programs.
- Adds custom testing mixins that provide extensive domain model data setup and convenient test assertions to standardize testing the API in light of permissions
- Extensively tests CRUD operations for each role type, plus superusers (and in some places, authed users and contributor users) for most API resources.

#### Why are we doing this? How does it help us?

Documented extensively in the related JIRA ticket

#### How should this be manually tested?

In test environment, go to Django admin. Assuming it's already populated with org/rp/page etc. data, create different users with different role types, and interact with the API via postman, authenticating as each distinct user.

In theory, this can also be tested by logging in to the SPA as, say, a revenue program admin and looking at the contributions list. You should only see contributions listed for your rps.  This pattern is true for any of the resources listed in the SPA admin.

#### Have automated unit tests been added?

- Adds custom testing mixins that provide extensive domain model data setup and convenient test assertions to standardize testing the API in light of permissions
- Extensively tests CRUD operations for each role type, plus superusers (and in some places, authed users and contributor users) for most API resources.
- Adds some tests in otherwise untouched code in order to get coverage > 95%


#### How should this change be communicated to end users?

NA


#### Are there any smells or added technical debt to note?

NA

#### Has this been documented? If so, where?

#### What are the relevant tickets?
**(Note: Please use syntax [JIRA-123] to link any relevant tickets)**

[DEV-699]

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

#### Are there next steps? If so, what? Have tickets been opened for them?


**(Note: Please use syntax [JIRA-123] to link any relevant tickets)**
